ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.utils
+
+import java.util.concurrent.Executors
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, Promise}
+
+trait ConcurrencyLimiter[T] {
+  def execute(task: () => Future[T]): Future[T]
+}
+
+class NoConcurrencyLimiter[T] extends ConcurrencyLimiter[T] {
+  override def execute(task: () => Future[T]): Future[T] = task()
+}
+
+class QueueBasedConcurrencyLimiter[T](
+    parallelism: Int,
+    executionContext: ExecutionContext,
+) extends ConcurrencyLimiter[T] {
+  assert(parallelism > 0)
+
+  case class WaitingTask(
+      task: () => Future[T],
+      promise: Promise[T] = Promise[T](),
+  )
+  private val waiting = mutable.Queue[WaitingTask]()
+  private var running: Int = 0
+
+  override def execute(task: () => Future[T]): Future[T] = synchronized {
+    if (running >= parallelism) {
+      val waitingTask = WaitingTask(task)
+      waiting.enqueue(waitingTask)
+      waitingTask.promise.future
+    } else {
+      startTask(task())
```

Comment:
I would consider not doing this optimization with the enqueue/start, and just having one path with enqueueing, which would make the code lighter